src/3d/symbols/qgspolygon3dsymbol_p.cpp

```diff
@@ -262,28 +260,81 @@ static Qt3DRender::QCullFace::CullingMode _qt3DcullingMode( Qgs3DTypes::CullingM
   return Qt3DRender::QCullFace::NoCulling;
 }
 
-Qt3DExtras::QPhongMaterial *QgsPolygon3DSymbolHandler::material( const QgsPolygon3DSymbol &symbol ) const
+Qt3DRender::QMaterial *QgsPolygon3DSymbolHandler::material( const QgsPolygon3DSymbol &symbol, bool isSelected, const Qgs3DRenderContext &context ) const
 {
-  Qt3DExtras::QPhongMaterial *material = new Qt3DExtras::QPhongMaterial;
+  Qt3DRender::QMaterial *retMaterial = nullptr;
+  if ( symbol.material().isUsingDiffuseTexture() )
+  {
+    QString textureFilePath = symbol.material().texturePath();
+    Qt3DExtras::QDiffuseMapMaterial *material = new Qt3DExtras::QDiffuseMapMaterial;
+
+    // front/back side culling
+    auto techniques = material->effect()->techniques();
+    for ( auto tit = techniques.constBegin(); tit != techniques.constEnd(); ++tit )
+    {
+      auto renderPasses = ( *tit )->renderPasses();
+      for ( auto rpit = renderPasses.begin(); rpit != renderPasses.end(); ++rpit )
+      {
+        Qt3DRender::QCullFace *cullFace = new Qt3DRender::QCullFace;
+        cullFace->setMode( _qt3DcullingMode( symbol.cullingMode() ) );
+        ( *rpit )->addRenderState( cullFace );
+      }
+    }
+    Qt3DRender::QTextureImage *textureImage = new Qt3DRender::QTextureImage;
+    textureImage->setSource( QUrl::fromLocalFile( textureFilePath ) );
```

Comment:
Looking at https://stackoverflow.com/questions/50944249/create-qtextureimage-from-qimage , it seems there's a means to directly construct textures from a QImage object. If this is the case, then you could use `QgsImageCache::pathAsImage() ` (via `QgsApplication::imageCache()->pathAsImage(...)`) here instead of directly loading from a local file.

There's two benefits to this approach

1. If the image has been previously loaded then it'll save a bunch of work
2. QgsImageCache supports base64 encoded image paths. This allows for embedding images directly inside a QgsProject, so that the project becomes portable and there's no path issues when you share the project (or QML style) with others. To enable this you'd also need to chance the `textureFile` widget in the ui file to a `QgsImageSourceLineEdit` widget (which adds the UI for users to embed pictures).

Reply:
I didn't know about that. I will change the textureFile widget to QgsImageSourceLineEdit and use QgsImageCache and make another commit.
Thanks for the suggestion!